python/lsst/daf/butler/server.py

```diff
@@ -0,0 +1,441 @@
+) -> list[str]:
+
+    expression_params = ExpressionQueryParameter(regex=regex, glob=glob)
+    collectionTypes = CollectionType.from_names(collectionType)
+    dataset_type = butler.registry.getDatasetType(datasetType) if datasetType else None
+
+    collections = butler.registry.queryCollections(
+        expression=expression_params.expression(),
+        datasetType=dataset_type,
+        collectionTypes=collectionTypes,
+        flattenChains=flattenChains,
+        includeChains=includeChains,
+    )
+    return list(collections)
+
+
+@app.get("/butler/v1/registry/collection/type/{name:path}", response_model=str)
+def get_collection_type(name: str, butler: Butler = Depends(butler_readonly_dependency)) -> str:
+    collectionType = butler.registry.getCollectionType(name)
+    return collectionType.name
+
+
+@app.put("/butler/v1/registry/collection/{name:path}/{type_}", response_model=str)
+def register_collection(
+    name: str,
+    collectionTypeName: CollectionTypeNames,
+    doc: str | None = Query(None),
+    butler: Butler = Depends(butler_readwrite_dependency),
+) -> str:
+    collectionType = CollectionType.from_name(collectionTypeName)
+    butler.registry.registerCollection(name, collectionType, doc)
+
+    # Need to refresh the global read only butler otherwise other clients
+    # may not see this change.
+    if GLOBAL_READONLY_BUTLER is not None:  # for mypy
+        GLOBAL_READONLY_BUTLER.registry.refresh()
+
+    return name
+
+
+@app.get(
+    "/butler/v1/registry/dataset/{id}",
+    summary="Retrieve this dataset definition.",
+    response_model=SerializedDatasetRef | None,
+    response_model_exclude_unset=True,
+    response_model_exclude_defaults=True,
+    response_model_exclude_none=True,
+)
+def get_dataset(
+    id: DatasetId, butler: Butler = Depends(butler_readonly_dependency)
+) -> SerializedDatasetRef | None:
+    ref = butler.registry.getDataset(id)
+    if ref is not None:
+        return ref.to_simple()
+    # This could raise a 404 since id is not found. The standard regsitry
+    # getDataset method returns without error so follow that example here.
+    return ref
+
+
+@app.get("/butler/v1/registry/datasetLocations/{id}", response_model=list[str])
+def get_dataset_locations(id: DatasetId, butler: Butler = Depends(butler_readonly_dependency)) -> list[str]:
+    # Takes an ID so need to convert to a real DatasetRef
+    fake_ref = SerializedDatasetRef(id=id)
+
+    try:
+        # Converting this to a real DatasetRef takes time and is not
+        # needed internally since only the ID is used.
+        ref = DatasetRef.from_simple(fake_ref, registry=butler.registry)
+    except Exception:
+        # SQL getDatasetLocations looks at ID in datastore and does not
+        # check it is in registry. Follow that example and return without
+        # error.
+        return []
+
+    return list(butler.registry.getDatasetLocations(ref))
+
+
+# TimeSpan not yet a pydantic model
+@app.post(
```

Comment:
Should it be `GET`?

Reply:
I think it has to be a POST because of the Query parameter in the collections. It's been a while since I had to think about this. If you run the server up it has some very nice automatically created documentation that gets built from knowledge of all the types and models.